src/main/kotlin/nl/info/zac/zaak/ZaakService.kt

```diff
@@ -161,6 +169,98 @@ class ZaakService @Inject constructor(
+            ?.betrokkeneIdentificatie?.identificatie
+        val (user, userUpdated) = assignUser(zaak, user, reason, behandelaar)
+        val (group, groupUpdated) = assignGroup(zaak, groupId, reason)
+        changeZaakDataAssignment(zaak.uuid, group, user)
+        if (userUpdated || groupUpdated) {
+            indexingService.indexeerDirect(zaak.uuid.toString(), ZoekObjectType.ZAAK, false)
+        }
+    }
+
+    private fun assignUser(
+        zaak: Zaak,
+        userName: String?,
+        reason: String?,
+        behandelaar: String?
+    ): Pair<User?, kotlin.Boolean> {
+        var user: User? = null
+        var userUpdated = false
+        userName?.takeIf { it.isNotEmpty() }?.let {
+            user = identityService.readUser(it)
+            if (behandelaar != userName) {
+                zrcClientService.updateRol(
+                    zaak,
+                    bepaalRolMedewerker(user, zaak),
+                    reason
+                )
+                userUpdated = true
+            }
+        } ?: zrcClientService.deleteRol(zaak, BetrokkeneTypeEnum.MEDEWERKER, reason).also {
+            userUpdated = true
+        }
+        return Pair(user, userUpdated)
+    }
+
+    private fun assignGroup(
+        zaak: Zaak,
+        groupId: String,
+        reason: String?
+    ): Pair<Group, kotlin.Boolean> {
+        var groupUpdated = false
+        val group = identityService.readGroup(groupId)
+        zgwApiService.findGroepForZaak(zaak)?.betrokkeneIdentificatie?.identificatie.let { currentGroupId ->
+            // if the zaak is not already assigned to the requested group, assign it to this group
+            if (currentGroupId == null || currentGroupId != groupId) {
+                val role = bepaalRolGroep(group, zaak)
+                zrcClientService.updateRol(zaak, role, reason)
+                groupUpdated = true
+            }
+        }
+        return Pair(group, groupUpdated)
+    }
+
+    private fun changeZaakDataAssignment(
+        zaakUuid: UUID,
+        group: Group,
+        user: User?
+    ) {
+        if (bpmnService.isZaakProcessDriven(zaakUuid)) {
+            zaakVariabelenService.setGroup(zaakUuid, group.name)
+            user?.let {
+                zaakVariabelenService.setUser(zaakUuid, it.getFullName())
+            } ?: zaakVariabelenService.removeUser(zaakUuid)
+        }
+    }
+
+    fun assignZaakToLoggedInUser(zaak: Zaak, groupId: String, userId: String, reason: String?) {
+        identityService.validateIfUserIsInGroup(userId, groupId)
+
+        val user = assignLoggedInUserToZaak(zaak, userId, reason)
+        val (group) = assignGroup(
+            zaak = zaak,
+            groupId = groupId,
+            reason = reason
+        )
+        changeZaakDataAssignment(zaak.uuid, group, user)
+        indexingService.indexeerDirect(zaak.uuid.toString(), ZoekObjectType.ZAAK, false)
+    }
+
+    private fun assignLoggedInUserToZaak(zaak: Zaak, userId: String, reason: String?) =
```

Comment:
```suggestion
    private fun assignUserToZaak(zaak: Zaak, userId: String, reason: String?) =
```
? I don't see any logic pertaining to a logged in user here?